### PR TITLE
Add strpos scalar string function. #15624

### LINKED ISSUE #15624: Add `strpos` scalar string function.
### Problem Statement

Add `strpos(string, substring) -> int` function, as it exists in PostgreSQL, see https://www.postgresql.org/docs/current/functions-string.html

Example:
`strpos('high', 'ig') -> 2`

`strpos('abracadabra', 'abra') -> 0`

This would improve PostgreSQL and in my use case, Tableau integration.

### Possible Solutions

_No response_

### Considered Alternatives

_No response_

### PR DESCRIPTION
**Pull request by DHRUV6029**
Closes #15624

Added strpos(string , substring) scalar string function to find first occurrence index of substring in string (1-based indexing compatible with PostGreSQL) returns 0 if substring is not found in string. 

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by BaurzhanSakhariev** [comment]
ok to test

File: .idea/misc.xml [hunk 1]
```diff
@@ -26,4 +26,4 @@
   <component name="SuppressionsComponent">
     <option name="suppComments" value="[]" />
   </component>
-</project>
\ No newline at end of file
+</project>
```
Comment (BaurzhanSakhariev) [inline .idea/misc.xml:29]: Please revert

File: docs/appendices/release-notes/5.7.0.rst [hunk 2]
```diff
@@ -133,3 +133,8 @@ External contributions
   operator as syntax sugar for the :ref:`scalar-power` scalar function. E.g.::
 
     SELECT 5 ^ 2;
+
+- `Dhruv Patel <https://github.com/DHRUV6029>`_ added support for the 'strpos(string, substring)'
+  operator for the :ref:`scalar-string` scalar function. E.g.::
+
+    SELECT strpos('crate' , 'ate');
```
Comment (BaurzhanSakhariev) [inline docs/appendices/release-notes/5.7.0.rst:140]:
```suggestion

- `Dhruv Patel <https://github.com/DHRUV6029>`_ added support for the
  :ref:`strpos <scalar-strpos>` scalar function.    
```

Comment (BaurzhanSakhariev) [reply docs/appendices/release-notes/5.7.0.rst:140]: Fixed reference and also dropepd the example, function name is pretty much self explaining.

File: docs/general/builtins/scalar-functions.rst [hunk 3]
```diff
@@ -797,6 +797,25 @@ Returns: ``text``
     +--------+
     SELECT 1 row in set (... sec)
 
+.. _scalar-strpos:
+
+``strpos(string, substring)``
+-----------------------------
+
+Finds the first occurence index of substring in a string If substring does not exists it returns 0.
```
Comment (BaurzhanSakhariev) [inline docs/general/builtins/scalar-functions.rst:805]:
```suggestion

Returns first starting 1-based index of the specified substring within string. 
If substring is not found, returns zero.

```

File: server/src/main/java/io/crate/expression/scalar/string/StringPositionFucntion.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import io.crate.data.Input;
+import io.crate.metadata.Functions;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.BoundSignature;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+
+public final class StringPositionFucntion extends Scalar<Integer , Object> {
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/expression/scalar/string/StringPositionFucntion.java:33]: let's use  `Scalar<Integer , String>` and avoid  casting in 
` var text = (String) args[0].value();`

File: server/src/test/java/io/crate/expression/scalar/string/StringPositionFunctionTest.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+public class StringPositionFunctionTest extends ScalarTestCase {
+
+    @Test
+    public void find_position_when_string_exists() {
+        assertEvaluate("strpos('crate', 'ate')", 3);
+    }
+
+    @Test
+    public void find_position_when_string_does_not_exists() {
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/expression/scalar/string/StringPositionFunctionTest.java:36]:
```suggestion
    @Test
    public void find_position_when_substring_not_found() {
```

File: server/src/test/java/io/crate/expression/scalar/string/StringPositionFunctionTest.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+import org.junit.Test;
+
+import io.crate.expression.scalar.ScalarTestCase;
+
+public class StringPositionFunctionTest extends ScalarTestCase {
+
+    @Test
+    public void find_position_when_string_exists() {
+        assertEvaluate("strpos('crate', 'ate')", 3);
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/expression/scalar/string/StringPositionFunctionTest.java:32]:
```suggestion
    @Test
    public void find_position_when_substring_found() {
        assertEvaluate("strpos('crate', 'ate')", 3);
```

File: server/src/test/java/io/crate/expression/scalar/string/StringPositionFunctionTest.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Test
+    public void find_position_when_string_does_not_exists() {
+        assertEvaluate("strpos('crate', 'db')", 0);
+    }
+
+    @Test
+    public void find_position_when_string_is_with_spaces() {
+        assertEvaluate("strpos('This is crate', 'is')", 3);
+    }
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/expression/scalar/string/StringPositionFunctionTest.java:43]: I don't think there is anything special about space. 
Maybe can instead rename this to `test_multiple_occurrences` which is an interesting case and is also tested here

File: docs/general/builtins/scalar-functions.rst [hunk 8]
```diff
@@ -797,6 +797,25 @@ Returns: ``text``
...
+::
+
+    cr> SELECT strpos('crate' , 'ate');
+    +---+
+    | 3 |
+    +---+
+    | 3 |
+    +---+
+    SELECT 1 row in set (0.012 sec)
+
```
Comment (BaurzhanSakhariev) [inline docs/general/builtins/scalar-functions.rst:818]:
```suggestion
    | 3 |
    +---+
    SELECT 1 row in set (... sec)

```

Comment (BaurzhanSakhariev) [reply docs/general/builtins/scalar-functions.rst:818]: We shouldn't use concrete numbers for execution time as it's deterministic and causes doc test failure.

**Comment by BaurzhanSakhariev** [comment]
Hi @DHRUV6029, thanks a lot for you contribution!

I added some minor comments.

**Comment by matriv** [comment]
@DHRUV6029 Thank you!

I'd like to additionally ask you to add a couple of tests for null values for both arguments.

**Comment by DHRUV6029** [comment]
I have resolved all the comments

**Comment by DHRUV6029** [comment]
Hi please review the changes

**Review by BaurzhanSakhariev — APPROVED** [review]
Thanks!

**Comment by BaurzhanSakhariev** [comment]
I will take a look at failures

File: docs/general/builtins/scalar-functions.rst [hunk 9]
```diff
@@ -797,6 +797,26 @@ Returns: ``text``
     +--------+
     SELECT 1 row in set (... sec)
 
+.. _scalar-strpos:
+
+``strpos(string, substring)``
+-----------------------------
+
+Returns first starting 1-based index of the specified substring within string.
+If substring is not found, returns zero.
```
Comment (matriv) [inline docs/general/builtins/scalar-functions.rst:806]: Please add the null handling sentence.

**Comment by BaurzhanSakhariev** [comment]
Hi @DHRUV6029, we are on feature freeze period, so this won't be merged after 5.7 release and will be part of CrateDB 5.8

Thanks again for your contribution!

**Comment by mfussenegger** [comment]
Thanks again for the PR. We've merged it via https://github.com/crate/crate/pull/15815 

(In case you intend to contribute again, please use a feature branch when creating a PR, it allows us to add fixups directly to the PR, in this case it wasn't possible because the PR was created from `master`)

**Pull request closed without merging**